Roadmap.md

```diff
@@ -243,3 +243,151 @@ Fetch API:
+  // with a nil error, except when p.byteLength == 0. Callers should treat a
+  // return of 0 and nil as indicating that nothing happened; in particular it
+  // does not indicate EOF.
+  //
+  // Implementations must not retain p.
+  async read(p: ArrayBufferView): Promise<ReadResult>;
+}
+
+// Writer is the interface that wraps the basic Write method.
+// https://golang.org/pkg/io/#Writer
+interface Writer {
+  // Write writes p.byteLength bytes from p to the underlying data stream. It
+  // returns the number of bytes written from p (0 <= n <= p.byteLength) and any
+  // error encountered that caused the write to stop early. Write must return a
+  // non-nil error if it returns n < p.byteLength. Write must not modify the
+  // slice data, even temporarily.
+  //
+  // Implementations must not retain p.
+  async write(p: ArrayBufferView): Promise<number>;
+}
+
+// https://golang.org/pkg/io/#Closer
+interface Closer {
+  // The behavior of Close after the first call is undefined. Specific
+  // implementations may document their own behavior.
+  close(): void;
+}
+
+// https://golang.org/pkg/io/#Seeker
+interface Seeker {
+  // Seek sets the offset for the next Read or Write to offset, interpreted
+  // according to whence: SeekStart means relative to the start of the file,
+  // SeekCurrent means relative to the current offset, and SeekEnd means
+  // relative to the end. Seek returns the new offset relative to the start of
+  // the file and an error, if any.
+  //
+  // Seeking to an offset before the start of the file is an error. Seeking to
+  // any positive offset is legal, but the behavior of subsequent I/O operations
+  // on the underlying object is implementation-dependent.
+  async seek(offset: number, whence: number): Promise<void>;
+}
+
+// https://golang.org/pkg/io/#ReadCloser
+interface ReaderCloser extends Reader, Closer { }
+
+// https://golang.org/pkg/io/#WriteCloser
+interface WriteCloser extends Writer, Closer { }
+
+// https://golang.org/pkg/io/#ReadSeeker
+interface ReadSeeker extends Reader, Seeker { }
+
+// https://golang.org/pkg/io/#WriteSeeker
+interface WriteSeeker extends Writer, Seeker { }
+
+// https://golang.org/pkg/io/#ReadWriteCloser
+interface ReadWriteCloser extends Reader, Writer, Closer { }
+
+// https://golang.org/pkg/io/#ReadWriteSeeker
+interface ReadWriteSeeker extends Reader, Writer, Seeker { }
+```
+These interfaces are well specified, simple, and have very nice utility
+functions that will be easy to port. Some example utilites:
+```ts
+// Copy copies from src to dst until either EOF is reached on src or an error
+// occurs. It returns the number of bytes copied and the first error encountered
+// while copying, if any.
+//
+// Because Copy is defined to read from src until EOF, it does not treat an EOF
+// from Read as an error to be reported.
+//
+// https://golang.org/pkg/io/#Copy
+async function copy(dst: Writer, src: Reader): Promise<number> {
```

Comment:
This could be optimized if there was an interface that the reader could implement to write directly into the writer. If a reader is managing internal buffers then this logic might force copies. .NET has a virtual method `CopyToAsync` on the `Stream` abstraction that allows for this https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=netcore-2.1

Reply:
It certainly would be nice to have [`sendfile(2)`](http://man7.org/linux/man-pages/man2/sendfile.2.html) support. 

I will leave adding a `copyTo` method for further revisions - as it's an optimization that won't exist in general - but thanks for bringing it up. (I'm not sure how/if Go does this...)

Reply:
Here it is https://github.com/golang/go/blob/15ac56fd60a28879fd65f6d761cc30c4c57ec0f1/src/io/io.go#L181.

> // WriterTo is the interface that wraps the WriteTo method.
> //
> // WriteTo writes data to w until there's no more data to write or
> // when an error occurs. The return value n is the number of bytes
> // written. Any error encountered during the write is also returned.
> //
> // The Copy function uses WriterTo if available.